Corrections to the original Tokeniser System

Dkrenon opened the pull request:
-Added the Grammar Type to the enums and Tokeniser System.
-Corrected the variable on the verbData component
-On the Tokeniser System added also the returns types that will be used on the ProcessCommandsTokens. 
-Created the test Tokeniser. As mentioned on Discord : "fish tokens so the action "kick" is the VRB, the obj "ball" is the DOBJ and the dobj(direction object) "window" is the IOBJ . In the enums ObjectType is just for the type of object: ball, football, knife, etc
also the dirObjectType is for the direction object: window,  door, staits, etc.
- Added the game constans for errors and  other stuff.
- On main.go the NTokeniserSystem is registered on the systems. This allows other systems to access the functions there.

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -100,58 +107,98 @@ func (ts *TokeniserSystem) setupVrbAct() {
-				if !objExists {
-					err = 1 // Set error code if no object is found
+	if obj == enums.ObjectTypeNone && dobj == enums.DirObjectTypeNone {
+		data.ErrCode = constants.ER_TKPR_NO
```
I would make a set of error typ[es one for tokeniser one for parser etc and then set a human readable error. these error codes are because we have no error type in solidity we do in go

Dkrenon replied:
done

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -100,58 +107,98 @@ func (ts *TokeniserSystem) setupVrbAct() {
+		if obj != enums.ObjectTypeNone && len(tokens) <= 3 {
+			data.DirectNoun = obj
+		} else if obj == enums.ObjectTypeNone && len(tokens) <= 3 {
+			err = constants.ER_TKPR_NO
```
make an error type

Dkrenon replied:
done

lbdl commented on cardinal/constants/game_constants.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+// Define the constants for ErrCodes
+const (
```
these should probably in fact be errors see https://go.dev/blog/error-handling-and-go we use these constatnt because as usual Solidity. Try to be idiomatic in Go this is a Go project. The LOGIC of the Solidity version of the engine is important not the code. You can use the codes in an actual error so we will leave these in. I would give them more meaningful names. Either that or we need a handler that generates a decent message based on the code.

Dkrenon replied:
Done, create a set of errors that have a struct of code and message that will make easy to access them. I changed a the err variable on the verbData struct to match the errors

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -100,58 +107,98 @@ func (ts *TokeniserSystem) setupVrbAct() {
 func (ts *TokeniserSystem) FishTokens(tokens []string) component.VerbData {
 	var data component.VerbData // Initialize VerbData to store the result
-	var err uint8               // Error code variable
+	var err uint8 = 0           // Error code variable
```
see Go zero values https://go.dev/tour/basics/12 we don't need to add an explicit value the compiler does this for us

Dkrenon replied:
done

lbdl commented on cardinal/system/t_tokeniser_system_test.go:
```diff
@@ -0,0 +1,81 @@
+func TestNewTokeniserSystem(t *testing.T) {
+	ts := NewTokeniserSystem()
+
+	assert.NotNil(t, ts.cmdLookup)
```
these SHOULD return nil if no key is passed to them or rather a "None" type

Dkrenon replied:
This func test the setup only, the next function test if no key or key is passed. Also adjusted the return functions on the tokeniser system so that they return a "None" type if no key is passed.

lbdl commented on cardinal/system/t_tokeniser_system_test.go:
```diff
@@ -0,0 +1,81 @@
+func TestFishTokensKickTheBallToTheWindow(t *testing.T) {
+	ts := NewTokeniserSystem()
+
+	tokens := []string{"KICK", "THE", "BALL", "TO", "THE", "WINDOW"}
```
we dont need the Object type. we Need the VRB the DOBJ and the IOBJ if it exists

lbdl replied:
to put another way this should be VRB:kick, DOBJ:football, IOBJ:window there is no object for the tokeniser, ever only the 3 mentioned things, (vrb, dobj, iobj)

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -100,58 +107,98 @@ func (ts *TokeniserSystem) setupVrbAct() {
-		return data
-	}
+	var vrb enums.ActionType = ts.cmdLookup[(tokens[0])]
+	var obj enums.ObjectType = ts.objLookup[(tokens[lenTokens])]
```
this should be a DIRECTOBJECT

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -100,58 +107,98 @@ func (ts *TokeniserSystem) setupVrbAct() {
-	}
+	var vrb enums.ActionType = ts.cmdLookup[(tokens[0])]
+	var obj enums.ObjectType = ts.objLookup[(tokens[lenTokens])]
+	var dobj enums.DirObjectType = ts.dirObjLookup[(tokens[lenTokens])]
```
this should be an INDIRECTOBJECT, remember there is NO OBJECT it makes no sense to the tokeniser: the tokeniser looks for ACTION/VERB, DOBJ and IOBJ. kick the ball. v:kick, do:ball, io:none. "kick ball at window": v:kick, do:ball, io:window

Dkrenon replied:
done

lbdl reviewed (changes requested):
we have no concept of object at this stage, we just want a tuple of vrb, directobject and indirectobject. object makes no sense. 
the tokeniser is extracting data that will then be handled in other functions that extract further meaning.
also Go has first class errors, solidity doesn't. I would make some Go error's rather than use the solidity constants.

Dkrenon commented:
Added the set of error types and modified the Err type on the verbData to match the errors types. The errors have a struct with code (uint32) and message (string) this will allow access to them if needed. 
Changed the names of the variables so that they match correctly (this we talked on discord, it was a naming thing rather than referring it to an object.

lbdl commented on cardinal/component/component_verbData.go:
```diff
@@ -5,15 +5,18 @@ It enhances code readability, simplifies data passing, aids in error handling, a
-	IndirectObjNoun enums.ObjectType    // The indirect Object noun
-	ErrCode         uint8               // Error code for parsing the tokens
+	Verb            enums.ActionType      // The action/verb from the command
+	DirectNoun      enums.ObjectType      // The direct object of the action
```
I would rename to DirectObject

Dkrenon replied:
done

lbdl commented on cardinal/component/component_verbData.go:
```diff
@@ -5,15 +5,18 @@ It enhances code readability, simplifies data passing, aids in error handling, a
-	ErrCode         uint8               // Error code for parsing the tokens
+	Verb            enums.ActionType      // The action/verb from the command
+	DirectNoun      enums.ObjectType      // The direct object of the action
+	IndirectDirNoun enums.DirObjectType   // The indirect directional object of the action
```
rename to DirectionObject

Dkrenon replied:
done

lbdl commented on cardinal/component/component_verbData.go:
```diff
@@ -5,15 +5,18 @@ It enhances code readability, simplifies data passing, aids in error handling, a
+	Verb            enums.ActionType      // The action/verb from the command
+	DirectNoun      enums.ObjectType      // The direct object of the action
+	IndirectDirNoun enums.DirObjectType   // The indirect directional object of the action
+	IndirectObjNoun enums.ObjectType      // The indirect Object noun
```
rename to IndirectObject

Dkrenon replied:
done

lbdl commented on cardinal/system/system_tokeniser.go:
```diff
@@ -118,46 +118,45 @@ func (ts *TokeniserSystem) setupGrammar() {
-	var vrb enums.ActionType = ts.cmdLookup[(tokens[0])]
-	var obj enums.ObjectType = ts.objLookup[(tokens[lenTokens])]
-	var dobj enums.DirObjectType = ts.dirObjLookup[(tokens[lenTokens])]
+	var VRB enums.ActionType = ts.cmdLookup[(tokens[0])]
```
I think idiomaticaly it should be Vrb as the ones below

Dkrenon replied:
done

lbdl reviewed (changes requested):
yep this is good couple of minor naming things

Dkrenon commented:
Corrected names of variables

lbdl commented on cardinal/component/component_verbData.go:
```diff
@@ -13,8 +13,8 @@ import (
-	DirectNoun      enums.ObjectType      // The direct object of the action
-	IndirectDirNoun enums.DirObjectType   // The indirect directional object of the action
+	DirectObject    enums.ObjectType      // The direct object of the action
+	IndirectObject  enums.DirObjectType   // The indirect directional object of the action
```
sorry my bad, this is the DIRECTION right? so its should be DirectionObject

lbdl replied:
we want a DirectObject, and IndirectObject and I guess a DirectionObject

Dkrenon replied:
enums.DirObjectType  is the direcction object: door, window, stairs, etc

Dkrenon replied:
I also have enums.DirectionType that are: north, south, up, left, etc

Dkrenon replied:
So, do I change it? if so, which will be the name?

lbdl replied:
so then it should be called a DIRECTIONOBJECT (ignore the upper case) and an DIRECTOBJECT should be a DIRECTOBJECT and and INDIRECTOBJECT should be called that is what I mean!

lbdl reviewed (changes requested):
I am confused ! the tokeniser needs to return the tokens that the rest of the system needs.
Those are the verb the directObject and the indirectObject.
I guess we can also add in a DirectionObject as well

lbdl commented:
merged but we need to change as we discussed
